go/decoder.go

```diff
@@ -0,0 +1,597 @@
+					}
+					err = dec.innerDecodeC(newReflectValue(reflect.ValueOf(&subb)), subc[0])
+					if err != nil {
+						log.Printf("error decoding sub bytes")
+						return err
+					}
+					allsize += len(subb)
+					parts = append(parts, subb)
+				}
+			}
+		} else {
+			val := make([]byte, aux)
+			_, err = io.ReadFull(dec.rin, val)
+			if err != nil {
+				return err
+			}
+			// Don't really care about count, ReadFull will make it all or none and we can just fall out with whatever error
+			return rv.SetBytes(val)
+			/*if (rv.Kind() == reflect.Slice) && (rv.Type().Elem().Kind() == reflect.Uint8) {
+				rv.SetBytes(val)
+			} else {
+				return fmt.Errorf("cannot write []byte to k=%s %s", rv.Kind().String(), rv.Type().String())
+			}*/
+		}
+	} else if cborType == cborText {
+		return dec.decodeText(rv, cborInfo, aux)
+	} else if cborType == cborArray {
+		return dec.decodeArray(rv, cborInfo, aux)
+	} else if cborType == cborMap {
+		return dec.decodeMap(rv, cborInfo, aux)
+	} else if cborType == cborTag {
+		/*var innerOb interface{}*/
+		ic := []byte{0}
+		_, err = io.ReadFull(dec.rin, ic)
+		if err != nil {
+			return err
+		}
+		if aux == tagBignum {
+			bn, err := dec.decodeBignum(ic[0])
+			if err != nil {
+				return err
+			}
+			return rv.SetBignum(bn)
+		} else if aux == tagNegBignum {
+			bn, err := dec.decodeBignum(ic[0])
+			if err != nil {
+				return err
+			}
+			minusOne := big.NewInt(-1)
+			bnOut := &big.Int{}
+			bnOut.Sub(minusOne, bn)
+			return rv.SetBignum(bnOut)
+		} else if aux == tagDecimal {
+			log.Printf("TODO: directly read bytes into decimal")
+		} else if aux == tagBigfloat {
+			log.Printf("TODO: directly read bytes into bigfloat")
+		} else {
+			decoder := dec.TagDecoders[aux]
+			var target interface{}
+			var trv DecodeValue
+			var err error
+
+			trv, target, err = rv.CreateTag(aux, decoder)
+			if err != nil {
+				return err
+			}
+
+			err = dec.innerDecodeC(trv, ic[0])
+			if err != nil {
+				return err
+			}
+
+			return rv.SetTag(aux, trv, decoder, target)
+		}
+		return nil
+	} else if cborType == cbor7 {
+		if cborInfo == int16Follows {
+			exp := (aux >> 10) & 0x01f
+			mant := aux & 0x03ff
+			var val float64
+			if exp == 0 {
+				val = math.Ldexp(float64(mant), -24)
+			} else if exp != 31 {
+				val = math.Ldexp(float64(mant+1024), int(exp-25))
+			} else if mant == 0 {
+				val = math.Inf(1)
+			} else {
+				val = math.NaN()
+			}
+			if (aux & 0x08000) != 0 {
+				val = -val
+			}
+			return rv.SetFloat64(val)
+		} else if cborInfo == int32Follows {
+			f := math.Float32frombits(uint32(aux))
+			return rv.SetFloat32(f)
+		} else if cborInfo == int64Follows {
+			d := math.Float64frombits(aux)
+			return rv.SetFloat64(d)
+		} else if cborInfo == cborFalse {
+			return rv.SetBool(false)
+		} else if cborInfo == cborTrue {
+			return rv.SetBool(true)
+		} else if cborInfo == cborNull {
+			return rv.SetNil()
+		}
```

Comment:
when switching to a switch statement, make sure to have a default case to catch errors